### PR TITLE
[Merged by Bors] - feat: port Data.Set.Basic

### PR DESCRIPTION
**Pull request by kim-em**
1b36dabc50929b36caec16306358a5cc44ab441e

If more PRs arrive in mathlib that slice up Data.Set.Basic, that's okay, we'll just delete the relevant sections here.

TODO: finish linting

### CODE REVIEW

File: Mathlib/Data/Set/Basic.lean [hunk 1]
```diff
@@ -0,0 +1,2740 @@
...
+  Also works for `{p : Prop} (f : p → α)` (unlike `image`)
+
+* `inclusion s₁ s₂ : ↥s₁ → ↥s₂` : the map `↥s₁ → ↥s₂` induced by an inclusion `s₁ ⊆ s₂`.
+
+## Notation
+
+* `f ⁻¹' t` for `preimage f t`
+
+* `f !! s` for `image f s`
+
```
Comment (eric-wieser) [inline Mathlib/Data/Set/Basic.lean:58]: This docstring is out of sync
```suggestion
```

Comment (eric-wieser) [reply Mathlib/Data/Set/Basic.lean:58]: (in lots of places actually; can you just grab it afresh from mathlib3?

**Comment by ChrisHughes24** [comment]
We have to make a decision on `simp` normal forms. Do we remove the `simp` attribute from `nonempty_subtype {α} {p : α → Prop} : Nonempty (Subtype p) ↔ ∃ a : α, p a` or  `nonempty_coe_sort {s : Set α} : Nonempty (↥s) ↔ s.Nonempty`. There are other similar examples. I propose keeping the `simp` attribute on the `set` lemmas

File: Mathlib/Data/Set/Basic.lean [hunk 2]
```diff
@@ -0,0 +1,2749 @@
...
+    else
+      simp [h]
+#align set.univ_eq_true_false Set.univ_eq_true_false
+
+section Preorder
+
+variable [Preorder α] [Preorder β] {f : α → β}
+
+theorem monotoneOn_iff_monotone : MonotoneOn f s ↔
+  (@Monotone _ _ (Subtype.instPreorderSubtype _) _ fun a : s => f a) := by
```
Comment (winstonyin) [inline Mathlib/Data/Set/Basic.lean:2519]: Bandaid fix for broken instance search, here and below.

**Comment by Ruben-VandeVelde** [comment]
Lint failures:
```lean
./build/bin/runLinter
-- Found 2 errors in 15315 declarations (plus 44200 automatically generated ones) in mathlib with 13 linters

/- The `simpNF` linter reports:
SOME SIMP LEMMAS ARE NOT IN SIMP-NORMAL FORM.
see note [simp-normal form] for tips how to debug this.
https://leanprover-community.github.io/mathlib_docs/notes.html#simp-normal%20form -/
-- Mathlib.Data.Set.Basic
#check Set.sep_union.{u} /- Left-hand side simplifies from
  { x | x ∈ s ∪ t ∧ p x }
to
  { x | (x ∈ s ∨ x ∈ t) ∧ p x }
using
  simp only [Set.mem_union]
Try to change the left-hand side to the simplified term!
 -/
#check Set.sep_inter.{u} /- Left-hand side simplifies from
  { x | x ∈ s ∩ t ∧ p x }
to
  { x | (x ∈ s ∧ x ∈ t) ∧ p x }
using
  simp only [Set.mem_inter_iff]
Try to change the left-hand side to the simplified term!
 -/
```

**Comment by eric-wieser** [comment]
It looks like this is because `set.sep` is now unfolded syntactically like `coe`?

File: Mathlib/Data/Set/Basic.lean [hunk 3]
```diff
@@ -0,0 +1,2773 @@
...
+  let ⟨x, _, y, _, hxy⟩ := hs
+  ⟨⟨x, y, hxy⟩⟩
+#align set.nontrivial_of_nontrivial Set.nontrivial_of_nontrivial
+
+-- Porting note: simp_rw broken here
+/-- `s`, coerced to a type, is a nontrivial type if and only if `s` is a nontrivial set. -/
+@[simp, norm_cast]
+theorem nontrivial_coe_sort {s : Set α} : Nontrivial s ↔ s.Nontrivial := by
+  -- simp_rw [← nontrivial_univ_iff, Set.Nontrivial, mem_univ, exists_true_left, SetCoe.exists,
+  --   Subtype.mk_eq_mk]
```
[the remark this answers is not in the document]
Comment (kbuzzard) [reply Mathlib/Data/Set/Basic.lean:2476]: People are aware about this issue with Exists, which was caused by the introduction of the ` ↦` notation; at the porting meeting the idea was discussed about reverting the `↦` notation change or doing it "properly" (which would involve changes in core Lean).

**Review by Ruben-VandeVelde — APPROVED** [review]
I think this can probably land now

File: Mathlib/Data/Set/Basic.lean (hunk 2 above)
Comment (semorrison) [reply Mathlib/Data/Set/Basic.lean]: I've made `Elem` (previously `type_of_set` or something) reducible, so this instance appears, and restored the original statement. See https://leanprover.zulipchat.com/#narrow/stream/287929-mathlib4/topic/data.2Eset.2Ebasic/near/315272101

**Comment by semorrison** [comment]
bors merge

**Comment by semorrison** [comment]
bors merge

**Pull request closed without merging**